ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,524 @@
+	for grpIdx := range groupsReady {
+		gr := &groupsReady[grpIdx]
+		gr.groupTaskIndex = grpIdx
+		gr.incompleteCount = int64(gr.balancesCount + gr.resourcesCount)
+		gr.balances = allBalances[usedBalances : usedBalances+gr.balancesCount]
+		if gr.resourcesCount > 0 {
+			gr.resources = allResources[usedResources : usedResources+gr.resourcesCount]
+			usedResources += gr.resourcesCount
+		}
+		usedBalances += gr.balancesCount
+		if gr.incompleteCount == 0 {
+			gr.incompleteCount = dependencyFreeGroup
+		}
+	}
+
+	taskIdx := int64(-1)
+	defer atomic.StoreInt64(&taskIdx, tasksCount)
+	// create few go-routines to load asyncroniously the account data.
+	for i := 0; i < asyncAccountLoadingThreadCount; i++ {
+		go p.asyncPrefetchRoutine(&tasksQueue, &taskIdx, groupDoneCh)
+	}
+
+	// iterate on the transaction groups tasks. This array retains the original order.
+	completed := make(map[int]bool)
+	for i := 0; i < len(p.groups); {
+	wait:
+		incompleteCount := atomic.LoadInt64(&groupsReady[i].incompleteCount)
+		if incompleteCount > 0 || (incompleteCount != dependencyFreeGroup && !completed[i]) {
+			select {
+			case doneIdx := <-groupDoneCh:
+				if doneIdx < 0 {
+					doneIdx = -doneIdx
+					// if there is an error, report the error to the output channel.
+					p.outChan <- loadedTransactionGroup{
+						err: groupsReady[doneIdx].err,
+					}
+					return
+				}
+				if doneIdx > i {
+					// mark future txn as ready.
+					completed[doneIdx] = true
+					goto wait
+				} else if doneIdx < i {
+					// it was already processed.
+					goto wait
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+		next := i
+		for ; next < len(p.groups); next++ {
+			if !completed[next] {
+				if next > i {
+					i = next
+					goto wait
+				}
+				// next == i
+			}
+
+			delete(completed, next)
+
+			// if we had no error, write the result to the output channel.
+			// this write will not block since we preallocated enough space on the channel.
+			p.outChan <- loadedTransactionGroup{
+				accounts:  groupsReady[next].balances,
+				resources: groupsReady[next].resources,
+			}
+		}
+		// if we get to this point, it means that we have no more transaction to process.
+		break
+	}
+}
+
+func (gt *groupTask) markCompletionAcct(idx int, br loadedAccountDataEntry, groupDoneCh chan int) {
+	gt.balances[idx] = br
+	if atomic.AddInt64(&gt.incompleteCount, -1) == 0 {
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionResource(idx int, res loadedResourcesEntry, groupDoneCh chan int) {
+	gt.resources[idx] = res
+	if atomic.AddInt64(&gt.incompleteCount, -1) == 0 {
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionAcctError(err error, groupDoneCh chan int) {
+	for {
+		curVal := atomic.LoadInt64(&gt.incompleteCount)
+		if curVal <= 0 {
+			return
+		}
+		if atomic.CompareAndSwapInt64(&gt.incompleteCount, curVal, 0) {
+			gt.err = err
+			groupDoneCh <- -gt.groupTaskIndex
```

Comment:
this is probably a good idea for performance but not for groupTaskIndex=0.
Maybe it is better returning a struct with an error flag and group index

Reply:
ahh - nice catch. while it's very tempting to just pack it all in a single structure and place it on the channel, it would reduce the efficiency of the channel. I think that by offsetting the index, we can easily dodge the index 0 issue. I'll document that so that it would be clear.